sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/transforms/CastTest.java

```diff
@@ -36,137 +32,252 @@
         pipeline
-            .apply(Create.of(new Nullable2()))
-            .apply(Cast.narrowing(outputSchema))
-            .apply(Convert.to(Nullable1.class));
+            .apply(Create.of(input).withRowSchema(inputSchema))
+            .apply(Cast.narrowing(outputSchema));
+
+    PAssert.that(output).containsInAnyOrder(expected);
 
-    PAssert.that(pojos).containsInAnyOrder(new Nullable1());
     pipeline.run();
   }
 
-  @Test(expected = IllegalArgumentException.class)
-  @Category(NeedsRunner.class)
-  public void testIgnoreNullableFail() throws Exception {
-    // ignoring nullable is opposite of weakening
-    Schema inputSchema = pipeline.getSchemaRegistry().getSchema(Nullable2.class);
-    Schema outputSchema = pipeline.getSchemaRegistry().getSchema(Nullable1.class);
+  @Test
+  public void testIgnoreNullableFail() {
+    // the opposite of testWeakenNullable
+
+    Schema inputSchema = Schema.of(Schema.Field.nullable("f0", Schema.FieldType.INT32));
+
+    Schema outputSchema = Schema.of(Schema.Field.of("f0", Schema.FieldType.INT64));
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage(
+        containsString("f0: Can't cast nullable field to non-nullable field"));
 
     Cast.widening(outputSchema).verifyCompatibility(inputSchema);
   }
 
   @Test
   @Category(NeedsRunner.class)
-  public void testComplexCast() throws Exception {
-    Schema outputSchema = pipeline.getSchemaRegistry().getSchema(All2.class);
-
-    PCollection<All2> pojos =
+  public void testCastInnerRow() {
+    Schema innerInputSchema =
+        Schema.of(
+            Schema.Field.of("f0", Schema.FieldType.INT16),
+            Schema.Field.of("f1", Schema.FieldType.INT32));
+
+    Schema inputSchema =
+        Schema.of(
+            Schema.Field.of("f0", Schema.FieldType.row(innerInputSchema)),
+            Schema.Field.of("f1", Schema.FieldType.INT32));
+
+    Schema innerOutputSchema =
+        Schema.of(
+            Schema.Field.of("f0", Schema.FieldType.INT32),
+            Schema.Field.of("f1", Schema.FieldType.INT64));
+
+    Schema outputSchema =
+        Schema.of(
+            Schema.Field.of("f0", Schema.FieldType.row(innerOutputSchema)),
+            Schema.Field.of("f1", Schema.FieldType.INT64));
+
+    Row input =
+        Row.withSchema(inputSchema)
+            .addValue(Row.withSchema(innerInputSchema).addValues((short) 1, 2).build())
+            .addValue(42)
+            .build();
+
+    Row expected =
+        Row.withSchema(outputSchema)
+            .addValue(Row.withSchema(innerOutputSchema).addValues(1, 2L).build())
+            .addValue(42L)
+            .build();
+
+    PCollection<Row> output =
         pipeline
-            .apply(Create.of(new All1()))
-            .apply(Cast.narrowing(outputSchema))
-            .apply(Convert.to(All2.class));
+            .apply(Create.of(input).withRowSchema(inputSchema))
+            .apply(Cast.narrowing(outputSchema));
```

Comment:
This is a widening cast, isn't it? Both of the inner schema's fields are widened.

Reply:
Good catch! You are right